worldedit-core/src/main/java/com/sk89q/worldedit/command/tool/BlockTool.java

```diff
@@ -37,5 +40,23 @@
      * @param clicked
      * @return true to cancel the original event which triggered this action (if possible)
      */
-    boolean actPrimary(Platform server, LocalConfiguration config, Player player, LocalSession session, Location clicked);
+    @Deprecated
+    default boolean actPrimary(Platform server, LocalConfiguration config, Player player, LocalSession session, Location clicked) {
+        throw new AssertionError("actPrimary must be overridden");
+    }
+
+    /**
+     * Perform the primary action of this tool.
+     *
+     * @param server The platform
+     * @param config The config instance
+     * @param player The player
+     * @param session The local session
+     * @param clicked The location that was clicked
+     * @param face The face that was clicked
+     * @return true to cancel the original event which triggered this action (if possible)
+     */
+    default boolean actPrimary(Platform server, LocalConfiguration config, Player player, LocalSession session, Location clicked, @Nullable Direction face) {
```

Comment:
Can we have some Javadoc added here to indicate that this will not be `default` after WE 8?